third_party/amd/lib/TritonAMDGPUTransforms/StreamPipeline.cpp

```diff
@@ -749,10 +761,11 @@ struct PipelinePass : impl::TritonAMDGPUStreamPipelineBase<PipelinePass> {
         continue;
       }
       (void)pipelineLoop(forOp, tt::getNumStagesOrDefault(forOp, numStages),
-                         globalPrefetch, localPrefetch, useAsyncCopy);
+                         globalPrefetch, localPrefetch, useAsyncCopy,
+                         waitAtTail);
     }
 
-    if (useAsyncCopy) {
+    if (useAsyncCopy && !waitAtTail) {
```

Comment:
If we have 2 loops and one is not pingpong'ed we will make it slower because we now have a barrier for each load. We should change `combineRedundantWaitOps` to apply to a forOp body so we can call it inside the loop.

Reply:
The problem is currently there's a bug in combineRedundantWaitOps which incorrectly places the ops and crashes with num_stages=3. I think the performance loss from the separated wait on non-pingpong schedule is not too significant.
This will be fixed when the bug is fixed.